qa/cephfs: add tests for ACLs

rishabh-d-dave opened the pull request:
This PR intends to add test for ACLs on CephFS mounted via kernel and FUSE.

rishabh-d-dave commented:
* added more tests
* rewrote some tests
* added a method to run a command as a certain user
* added a option to allow sudo in command arguments (since `sudo -u` might be used by some).
* added an parameter to run_shell() to allow skipping execution abortion when supplied command fails
* tested to check if all test written so far are working fine
* added more TODOs

rishabh-d-dave commented:
changelog:
changes in tests -
* implemented stub methods `test_acls_on_current_user` and `test_acls_on_current_group`
* added 6 more tests (number of test increased from 8 to 14)
* rewrote all tests accordingy to following changes

changes outside tests -
* added method to create a file
    * provided a default value to parameter for file name
    * added the ability to create the file as different user
* added the ability to add input to `run_shell`'s stdin
* added the abililty to accept commands as strings
* added a method to run given command as root user
* added a check to make sure that commands to be ran as a diferent user does not
have quotes around them. (Python's shell simulation mistakes it as the part of command)
* made improvements to the code that allows sudo in vstart_runner's `run()`

rishabh-d-dave commented:
Except for `test_acls_on_other_user` and `test_acls_on_other_group` all tests are passing. The acl `user:dummy:rwx` and `group:dummy:rwx` has no  effect -

```
rishabh$ ll /tmp/tmpFKmknT/mnt.0/
total 1
drwxrwxr-x+ 2 rishabh rishabh 0 Feb 26 12:27 dir1
rishabh$ getfacl /tmp/tmpFKmknT/mnt.0/dir1/
getfacl: Removing leading '/' from absolute path names
# file: tmp/tmpFKmknT/mnt.0/dir1/
# owner: rishabh
# group: rishabh
user::rwx
user:dummy:rwx
group::rwx
mask::rwx
other::r-x

rishabh$ sudo -u dummy -s /bin/bash -c "ls /tmp/tmpFKmknT/mnt.0/dir1/"
ls: cannot access '/tmp/tmpFKmknT/mnt.0/dir1/': Permission denied
rishabh$
```

The same ACL works on non ceph FS -
```
# works on my FS
rishabh$ ll | grep dir1
drwxrwxr-x+ 2 rishabh rishabh 4096 Feb 26 11:23 dir1
rishabh$ ls dir1
$ getfacl dir1
# file: dir1
# owner: rishabh
# group: rishabh
user::rwx
user:dummy:rwx
group::rwx
mask::rwx
other::r-x

rishabh$ sudo mkdir dir1/dir2
rishabh$ sudo -u dummy -s /bin/bash -c "ls dir1"
dir2
rishabh$
```

@batrick 
Is what you said in your last mail relevant to this? I am a bit confused since `test_acls_on_current_user` and `test_acls_on_current_group` passed.

rishabh-d-dave commented:
>  If you just wanted to test libcephfs, you could have done this by passing in UserCred objects with the correct info instead. That said, this design should allow this to run on kcephfs and fuse, so maybe this is better.

Since the tracker issue was from FUSE POV, I never thought about testing libcephfs directly. Are there any serious benefits of that approach?

rishabh-d-dave commented:
Also I must admit that I could have written tests more elegantly. I will make these non-functional changes a little later so that I can make some progress on my other PRs too.

rishabh-d-dave commented:
I've been staring at the screen for a while 'cause I thought I was missing something that would allow access to user `dummy` for `/tmp/tmpNNtT8m/mnt.0/dir1`. `dummy` should've been able to access it, right?

```
rishabh$ getfacl /tmp/tmpNNtT8m/mnt.0/dir1
getfacl: Removing leading '/' from absolute path names
# file: tmp/tmpNNtT8m/mnt.0/dir1
# owner: root
# group: root
user::rwx
user:rishabh:rwx
user:dummy:rwx
group::rwx
group:rishabh:rwx
mask::rwx
other::rwx
default:user::rwx
default:user:rishabh:rwx
default:group::rwx
default:group:rishabh:rwx
default:mask::rwx
default:other::rwx

rishabh$ sudo -u dummy -s /bin/bash -c "ls /tmp/tmpNNtT8m/mnt.0/dir1"
ls: cannot access '/tmp/tmpNNtT8m/mnt.0/dir1': Permission denied
rishabh
```

Then I thought this is a bug since user classified as "other" worked ext4 -

```
rishabh$ sudo setfacl -m o:rwx dir1
rishabh$ sudo -u dummy -s /bin/bash -c "ls dir1"
dir2
rishabh$ getfacl dir1
# file: dir1
# owner: root
# group: root
user::rwx
group::rwx
other::rwx
default:user::rwx
default:group::rwx
default:other::rwx
```

But as I write I recall Patrick's reply on negative tests  mail thread. `dummy` can't access CephFS mount because it's a Linux user and not a Ceph user, right? I can see similar behaviour [here](https://github.com/ceph/ceph/pull/26477#issuecomment-467344248)

rishabh-d-dave commented:
changes:
* made several changes to reduce total LOC
   * total LOC in test_acls.py reduced by around 250 lines
   * moved all the similar looking code to some a new method with minimum branching
   * defined all LOCs collectively in a separate method

* all tests test `r`, `w` as well as `x` permissions
* added a preliminary test wherever possible

jtlayton commented:
> Since the tracker issue was from FUSE POV, I never thought about testing libcephfs directly. Are there any serious benefits of that approach?

It's generally faster since you aren't going through the kernel. It also doesn't require any sort of elevated (kernel) capabilities.

But...being able to run the same tests with kcephfs is genuinely useful, so I think that overrides any benefit from doing this vs. libcephfs directly.

jtlayton commented on qa/tasks/cephfs/mount.py:
```diff
@@ -142,6 +142,11 @@ def create_files(self):
+                         pwd=None, check_status=True):
+        return self.create_file(filename=filename, dirname=dirname, user=user,
+                         pwd=pwd, check_status=False)
+
```
The subject on this patch should probably be "add a method to test if file can be created". This doesn't actually write to the file, AFAICT.

rishabh-d-dave replied:
You are right.

jtlayton commented:
This looks reasonable overall.

Just to play devil's advocate...is this something that we need to maintain ourselves? I have to wonder if there are standard POSIX ACL testsuites out there we should be using here instead. I'll ask around...

jtlayton commented:
> rishabh$ sudo -u dummy -s /bin/bash -c "ls /tmp/tmpFKmknT/mnt.0/dir1/"
> ls: cannot access '/tmp/tmpFKmknT/mnt.0/dir1/': Permission denied
> rishabh$
> @batrick
> Is what you said in your last mail relevant to this? I am a bit confused since `test_acls_on_current_user` and `test_acls_on_current_group` passed.

I'd be more inclined to think that some other component of the path might be restricting permissions. Bear in mind that dummy needs read permissions on all directory inodes the path.

jtlayton commented:
Approved, but you will need to fix up the bogus submodule update here.

jtlayton commented on qa/tasks/cephfs/test_acls.py:
```diff
@@ -0,0 +1,411 @@
+    #
+    ####################################################
+
+    def test_options_Rx_for_all_acls(self):
```
I think the "recursive" tests here are really just testing the functionality of setfacl. setfacl will just walk the tree recursively and update the xattr of each object. I think you can probably drop most of these recursive tests.

rishabh-d-dave replied:
> I think the "recursive" tests here are really just testing the functionality of setfacl. setfacl will just walk the tree recursively and update the xattr of each object. I think you can probably drop most of these recursive tests.

I admit that I did feel this after I completed writing these tests. Originally, this was a single testcase (named `set, view and remove ACLs recursively`) among the rest, so I couldn't see that straightaway.

Yesterday, I had a thought on expanding the recursive tests for default ACLs. Right now, the testcase for default ACLs has the following steps:

1. Apply ACLs to the mountpoint.
2. Create directories inside
3. Check if default ACLs worked as expected on newly created directories.

Following could be a better case, if we want it to keep it in the first place -

1. Create directories
2. Apply ACLs to the mountpoint.
3. Check if default ACLs are on mountpoint and on the directories within it.
4. Create directories within directories
5. Check if default ACLs corresponding to default ACLs are present on newly created directories.

Is this worth doing/keeping? Or... this is just a complex version of [test_default_acls()](https://github.com/ceph/ceph/pull/26477/files#diff-c2417cd8e2afc1b19952425d5f25534fR254)? In case we want to keep these tests, the current approach of recursive tests isn't good enough, IMO.

> FWIW, I did ask around and didn't find any comprehensive testsuites for this. xfstests has some ACL related tests, and I saw a fork of pjdfstests that had some too. Maybe we should just look at leveraging those for this testing? Is that unreasonable?

I can't say that myself since I wasn't aware of these tests until now.
 
> In any case, I'm a little unclear on what the exact requirements are here. Are you looking to specifically test setting/removing ACLs? Or are you more interested in testing whether ACLs are being properly enforced for different POSIX filesystem operations (of which the setxattr that handles the ACL change is only one).

Actually both. When I started with this PR, I had the intention to write tests that would make sure that ACLs are working as one would expect them to work on ext4 no matter what underlying FS is used with CephFS.

jtlayton replied:
Yes, we'd definitely want to test ACL inheritance. That said, will testing beyond one level of inheritance be useful?

If we're interested in ensuring ACL compliance over kernel-based filesystems, might we be better served by adapting existing testsuites? I imagine they will be more comprehensive than anything we could come up with.

FWIW:

LTP has this: https://github.com/linux-test-project/ltp/tree/master/testcases/kernel/fs/acl

...pretty old, but would be reasonably easy to adapt to cephfs.

I looked closer at xfstests: https://git.kernel.org/pub/scm/fs/xfs/xfstests-dev.git

If you do a git grep in there for "setfacl", there are some tests that call it. Most of them seem to be regression tests for specific bugs though, so I'm not sure if they're really quite what we'd want.

This guy has a small suite of ACL tests that might be useful: https://github.com/Hnasar/pjdfstest/tree/master/tests/xacl

...but it's not clear to me why he didn't contribute them back to upstream pjdfstests though. They seem to be BSD licensed, so we should be able to use them for this if we wish (just be sure to attribute them properly).

rishabh-d-dave replied:
> Yes, we'd definitely want to test ACL inheritance. That said, will testing beyond one level of inheritance be useful?
>

Hmm, I get it. The new steps only adds complexity but the testcase remains the same. I'll go ahead and delete recursive test cases.

> If we're interested in ensuring ACL compliance over kernel-based filesystems, might we be better served by adapting existing testsuites? I imagine they will be more comprehensive than anything we could come up with.
>
> FWIW:
> 
> LTP has this: https://github.com/linux-test-project/ltp/tree/master/testcases/kernel/fs/acl
> 
> ...pretty old, but would be reasonably easy to adapt to cephfs.
> 
> I looked closer at xfstests: https://git.kernel.org/pub/scm/fs/xfs/xfstests-dev.git
> 
> If you do a git grep in there for "setfacl", there are some tests that call it. Most of them seem to be regression tests for specific bugs though, so I'm not sure if they're really quite what we'd want.
> 
> This guy has a small suite of ACL tests that might be useful: https://github.com/Hnasar/pjdfstest/tree/master/tests/xacl
> 
> ...but it's not clear to me why he didn't contribute them back to upstream pjdfstests though. They seem to be BSD licensed, so we should be able to use them for this if we wish (just be sure to attribute them properly).

Ok. I'll take a look and create a list of testcases these testsuites cover so that we can compare them with ours. Do we want to translate these test cases to Python or would be prefer to use them as it is?

jtlayton replied:
> Ok. I'll take a look and create a list of testcases these testsuites cover so that we can compare them with ours. Do we want to translate these test cases to Python or would be prefer to use them as it is?

I think you get to make that call, since you're doing the work. :) It really comes down to what you think will be easiest to do and (more importantly) maintain over the long haul.

rishabh-d-dave replied:
I think Python's better from maintainability POV. Especially after the rewriting I did late last week and early this week, the code in test methods is nearly same as writing English.

@jtlayton @batrick @ukernel @vshankar 
In case there is a different opinion on this, please let me know.

rishabh-d-dave commented:
changes:
* changed the commit title
* made some minor changes
* did a rebase

rishabh-d-dave commented:
changes:
* dropped recursive functions
* `test_default_acl()` and `test_acl_mask()` into subcases based on file owner, file group, other and current user and group
* minor changes in helper methods and there are one or few more

rishabh-d-dave commented:
@batrick 
> There's a change to the submodule civetweb. Please rebase and undo that.

Done.

jtlayton commented on qa/tasks/cephfs/test_acls.py:
```diff
@@ -0,0 +1,480 @@
+# 3. Assign appropriate permissions to directories
+# 4. Apply ACL
+# 5. Check for their presence
+# 6. check for theri effect
```
nit: their

jtlayton commented on qa/tasks/cephfs/test_acls.py:
```diff
@@ -0,0 +1,480 @@
+        else:
+            raise RuntimeError('var "%s" (of type %s) should be either list or str.' % (var, str(type(var))))
+
+    def assert_acls_are_(self, state, acls, testdirs):
```
nit: trailing underscore on the name here is odd, maybe remove it?

rishabh-d-dave replied:
The trailing underscore makes sense when reading calls made to this method. Calls can be `assert_acl_are('present', ...` or `assert_acl_are('absent', ...`. Just `assert_acls_are` sounded alot incomplete to me.

If you think this doesn't make sense, we can drop it.

rishabh-d-dave replied:
Replaced the method name in upcoming push by `assert_acls_are()` and added `assert_acls_are_present()` and `assert_acls_are_absent()`. This should it make it more convenient to write.

jtlayton commented on qa/tasks/cephfs/test_acls.py:
```diff
@@ -0,0 +1,480 @@
+        self.assert_acls_are_('absent', self.g_acl, self.testdir1)
+        self.assert_dirs_are_inaccessible(self.testdir1)
+
+    # XXX: Always fails
```
Do we know why this fails? Where does it fall down?

rishabh-d-dave replied:
I get the "permission denied" error message -
```
2019-03-20 18:58:31,865.865 INFO:__main__:Running ['sudo', '-u', 'dummy', '-s', '/bin/bash', '-c', 'ls /tmp/tmpLsYBMs/mnt.0/dir1']
ls: cannot access '/tmp/tmpLsYBMs/mnt.0/dir1': Permission denied
> /home/rishabh/repos/ceph/enable-acls-by-default/qa/tasks/cephfs/test_acls.py(357)_test_acl_for_other_user()
-> assert proc.returncode == 0
(Pdb) p proc.returncode
2
(Pdb) p proc.stderr().getvalues()
*** AttributeError: AttributeError('StringIO instance has no __call__ method',)
(Pdb) p proc.stderr().getvalue()
*** AttributeError: AttributeError('StringIO instance has no __call__ method',)
(Pdb) p proc.stderr.getvalue()
"ls: cannot access '/tmp/tmpLsYBMs/mnt.0/dir1': Permission denied\n"
(Pdb) p output
'# file: tmp/tmpLsYBMs/mnt.0/dir1\n# owner: root\n# group: root\nuser::rwx\nuser:dummy:rwx\ngroup::---\nmask::rwx\nother::---\n\n'
(Pdb) 
```

 However, this is never the case with user using which I run the test -
```
rishabh$ ls /tmp/tmpLsYBMs/mnt.0/dir1
ls: cannot open directory '/tmp/tmpLsYBMs/mnt.0/dir1': Permission denied
rishabh$ sudo ls /tmp/tmpLsYBMs/mnt.0/dir1
dir2
rishabh$ sudo -u dummy -s /bin/bash -c "ls /tmp/tmpLsYBMs/mnt.0/dir1"
ls: cannot access '/tmp/tmpLsYBMs/mnt.0/dir1': Permission denied

dummy$ ls /tmp/tmpLsYBMs/mnt.0/dir1
ls: cannot access '/tmp/tmpLsYBMs/mnt.0/dir1': Permission denied
dummy$ sudo getfacl /tmp/tmpLsYBMs/mnt.0/dir1
getfacl: Removing leading '/' from absolute path names
# file: tmp/tmpLsYBMs/mnt.0/dir1
# owner: root
# group: root
user::rwx
user:dummy:rwx
group::---
mask::rwx
other::---
dummy$ logout

rishabh$ sudo setfacl -m user:rishabh:rwx /tmp/tmpLsYBMs/mnt.0/dir1
rishabh$ ls /tmp/tmpLsYBMs/mnt.0/dir1
dir2
```

I felt that the absence of keyring might be the reason, but that doesn't looks like the case - 
```
rishabh$ sudo cp ~/ceph/build/keyring /home/dummy/.ceph
rishabh$ sudo diff ~/ceph/build/keyring /home/dummy/.ceph/keyring
rishabh$ sudo -u dummy -s /bin/bash -c "ls /tmp/tmpLsYBMs/mnt.0/dir1"
ls: cannot access '/tmp/tmpLsYBMs/mnt.0/dir1': Permission denied
rishabh$ sudo su - dummy
dummy$ ls /tmp/tmpLsYBMs/mnt.0/dir1
ls: cannot access '/tmp/tmpLsYBMs/mnt.0/dir1': Permission denied
dummy$ 

```

rishabh-d-dave replied:
This happens not only for a user in some other group but also for users in the same group -

```
rishabh$ sudo gpasswd -a user2 user2
Adding user user2 to group user2
rishabh$ groups user2
user2 : user2 rishabh
rishabh$ sudo -u user2 -s /bin/bash -c "ls /tmp/tmpLsYBMs/mnt.0/dir1"
ls: cannot access '/tmp/tmpLsYBMs/mnt.0/dir1': Permission denied
rishabh$ sudo su - user2
user2$ ll /tmp/tmpLsYBMs/mnt.0/
ls: cannot access '/tmp/tmpLsYBMs/mnt.0/': Permission denied
```

rishabh-d-dave replied:
The state of permission bits on the directory containing the mountpoint seems to be the troublemaker -

```
$ ls -la /tmp/tmpL8d3v_/
total 1
drwx------.  3 rishabh rishabh   60 Mar 25 17:04 .
drwxrwxrwt. 33 root    root    1740 Mar 25 17:04 ..
drwxrwxrwt.  2 rishabh rishabh    0 Mar 25 17:04 mnt.0
```

Fixing that fixes the issue with "file group" tests too.

jtlayton replied:
Glad you figured it out.

jtlayton commented on qa/tasks/cephfs/mount.py:
```diff
@@ -120,6 +142,11 @@ def create_files(self):
                 'sudo', 'touch', os.path.join(self.mountpoint, suffix)
             ])
 
+    def test_create_file(self, filename='testfile', dirname=None, user=None,
```
We'll definitely want to test file creates, but we need to test the entire range of operations that require permissions checks. These are primarily done via the Client::inode_permission methods in libcephfs. The various filesystem operations call the may_* functions that call inode_permission with various MAY_* flags.

We'll want to ensure that all of the MAY_* flag combinations to inode_permission are tested. It doesn't need to be done in this first patchset, but we do need to ensure that the entire range of dir ops are tested -- creates, deletes, renames, link, unlink. We'll also need to ensure that opening files is tested in various O_* flag and ACL combinations.

That said, I'm fine with merging an initial set like this and then extending it in follow-on sets.

jtlayton commented:
Just to make sure we're on the same page -- @batrick mentioned when we were chatting that he would prefer that we just integrate the LTP tests for this, similarly to how the pjdfstests were integrated. I tend to agree as we'll be able to integrate new tests from LTP with less hassle that way.

@rishabh-d-dave , do you agree or are you still looking at translating these tests into python?

batrick commented:
I believe Rishabh plans to open a new PR that does just that. Leaving this one as-is in case we want to take parts of it.

rishabh-d-dave commented:
*new tests*
* FEA: add tests for default mask ACLs (I had a untested old version of these lying around in a separate patch for quite some time, just lumped them together since they are not covered in LTP's testsuite and so that I don't forget them later)

*changes in tests addded during previous push*
* BUG in `test_default_acls_for_current_user` and  `test_default_acls_for_current_group`
  corollaries - 
    * add check that removes inherited ACL checks if that access was only due to the inherited ACL.
    * double check the behaviour since default non-mask ACL testcases do  not have a preliminary case
* BUG: `test_acl_mask_for_user` and `test_acl_mask_for_other` -- apparently mask affected ACL_OTHER which is not expected behaviour.
* REFACTOR: test method test one ACL at a time, names should reflect that 
* BUG: `test_remove_all_acls` failed because `g::rwx` changes to `g::---` when  mask ACL is applied

*changes in helper methods:*
* FEA: added methods to check if given ACL for file owner/group, other or other user/group ACL
* BUG: not all default ACLs are removable by `-x`
* BUG: nondefault ACLs can match similar default ACL during str.find(); make sure that doesn't happen
* BUG:  `assert_dirs_are_inaccessible()` was passing arguments incorrectly to `assert_dirs_are_accessible()` found it when I was testing
        `test_acl_mask_for_file_owner` -- apparently mask affected ACL_USER which is not expected behaviour.
* BUG: rewrote `change_grp_perms` part in `create_test_dirs()` -- chown and chmod were being executed recusrsively on `mntpt/dir1/dir2/dir3` rather than `mntpt/dir1`
* BUG: remove `test_with_other_user` and add `other_user` in `assert_dirs_are_accessible()` -- this allows testing via a user other than `self.otheruser`.
* REFACTOR: `assert_acls_are_()` -> `test_acls_are()`, `test_acls_are_present()` and `test_acls_are_absent()`
* REFACTOR: `assert_dirs_are_accessible` - > `assert_dirs_are()`,
            `assert_dirs_are_accessible()`, and `assert_dirs_are_inaccessible()`
* REFACTOR: move repetitive code in `assert_dirs_are_accessible()` to a new function
* FEA: don't just crash on AssertionError; also, print the reason behind it
* BUG: in `is_fo_acl()`, `is_fg_acl()`, `is_o_acl()` - look for 2 colons after 'user'/'group'/'mask'

*change not in test_acls.py*
* BUG: in `testcmd_as_user()` -- `str(type(list)).find('list')` -> `str(type(args)).find('list')`

rishabh-d-dave commented:
@jtlayton 
I spent some time documenting my changes rather than just recalling and writing them at the end. Let me know if that helps reviewing.  :)

rishabh-d-dave commented:
@jtlayton @batrick 
IMHO, commits other than the one adding ACL tests can proceed to merge. What do you guys think? Are they useful from generic POV? Should I go ahead and create a separate PR for them so that those commits can be merged regardless of where ACL tests merge?

rishabh-d-dave commented:
Just to make sure that this doesn't slip unnoticed, the reason for which "other_user/group" tests and "file_group" tests were failing was the permission bits on the containing directory. I figured this [here](https://github.com/ceph/ceph/pull/26477#discussion_r268612479). I'll add the fix after we decide to unfreeze the PR.

rishabh-d-dave commented:
Created a separate PR for rest of the 9 commits (PR #27481). Now this PR depends on that one.

batrick commented:
@rishabh-d-dave status on this? rebase too please

rishabh-d-dave commented:
Closing this PR in favour of PR #29421.
